hcloud/action_waiter.go

```diff
@@ -0,0 +1,102 @@
+package hcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+)
+
+type ActionWaiter interface {
+	WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error)
+	WaitForActions(ctx context.Context, actions []*Action) ([]*Action, error)
+	WaitForAction(ctx context.Context, action *Action) (*Action, error)
+}
+
+var _ ActionWaiter = (*ActionClient)(nil)
+
+// WaitForActionsFunc implements ActionWaiter.
+func (c *ActionClient) WaitForActionsFunc(ctx context.Context, actions []*Action, handleUpdate func(update *Action)) ([]*Action, error) {
+	result := make([]*Action, 0, len(actions))
+
+	running := make(map[int64]struct{}, len(actions))
+	for _, action := range actions {
+		if action.Status == ActionStatusRunning {
+			running[action.ID] = struct{}{}
+		} else {
+			result = append(result, action)
+		}
+	}
+
+	retries := 0
+	for {
+		if len(running) == 0 {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(c.action.client.pollBackoffFunc(retries)):
+			retries++
+		}
+
+		opts := ActionListOpts{Sort: []string{"status", "id"}}
+		for actionID := range running {
+			opts.ID = append(opts.ID, actionID)
+		}
+
+		updates, err := c.AllWithOpts(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(updates) != len(running) {
+			// Some actions do not exist in the API, fail early and prevent an infinite loop.
```

Comment:
> Do you know how the actions proxy handles an unavailable backend? Does it just not return the action or is there an error in that case?

The actions proxy will return an internal server error if a service is not responding. So an action will "disappear" from the returned actions only if the ID is unknown.